File: ModularMonolithVsMicroservice/python/pulsecore/deinterleaver.py
"""PRI-based emitter track grouping -- Python port of
common/src/deinterleaver.cpp / common/include/deinterleaver.h.
"""
from pulsecore import pulse_pb2


class _Track:
    __slots__ = (
        "track_id",
        "pulse_count",
        "last_start_sample",
        "pri_sum_seconds",
        "pri_count",
        "peak_sum",
    )

    def __init__(self, track_id: int):
        self.track_id = track_id
        self.pulse_count = 0
        self.last_start_sample = 0
        self.pri_sum_seconds = 0.0
        self.pri_count = 0
        self.peak_sum = 0.0


class Deinterleaver:
    def __init__(self, sample_rate_hz: float, pri_tolerance_seconds: float):
        self._sample_rate_hz = sample_rate_hz
        self._pri_tolerance_seconds = pri_tolerance_seconds
        self._tracks: list[_Track] = []
        self._next_track_id = 1

    def process(
        self, batch: "pulse_pb2.PulseEventBatch", out: "pulse_pb2.DeinterleaveSummary"
    ) -> None:
        sample_rate_hz = self._sample_rate_hz
        pri_tolerance_seconds = self._pri_tolerance_seconds
        tracks = self._tracks
        next_track_id = self._next_track_id

        for event in batch.events:
            start_sample = event.start_sample
            peak_amplitude = event.peak_amplitude
            pulse_time = start_sample / sample_rate_hz

            # Prefer the closest track whose predicted next-pulse time
            # falls within tolerance; fall back to a track that only has
            # one pulse so far (no PRI to test against yet) if no
            # established track matches.
            best_match = None
            best_diff = pri_tolerance_seconds
            seed_match = None

            for track in tracks:
                if track.pri_count > 0:
                    last_time = track.last_start_sample / sample_rate_hz
                    predicted = last_time + (track.pri_sum_seconds / track.pri_count)
                    diff = abs(predicted - pulse_time)
                    if diff <= best_diff:
                        best_diff = diff
                        best_match = track
                elif track.pulse_count == 1 and seed_match is None:
                    seed_match = track

            target = best_match if best_match is not None else seed_match
            if target is None:
                target = _Track(next_track_id)
                next_track_id += 1
                tracks.append(target)

            if target.pulse_count > 0:
                last_time = target.last_start_sample / sample_rate_hz
                target.pri_sum_seconds += pulse_time - last_time
                target.pri_count += 1
            target.last_start_sample = start_sample
            target.peak_sum += peak_amplitude
            target.pulse_count += 1

        self._next_track_id = next_track_id

        out.Clear()
        for track in tracks:
            t = out.tracks.add()
            t.track_id = track.track_id
            t.pulse_count = track.pulse_count
            t.estimated_pri_seconds = (
                track.pri_sum_seconds / track.pri_count if track.pri_count > 0 else 0.0
            )
            t.mean_peak_amplitude = (
                track.peak_sum / track.pulse_count if track.pulse_count > 0 else 0.0
            )

File: ModularMonolithVsMicroservice/common/src/deinterleaver.cpp
#include "deinterleaver.h"

#include <cmath>

namespace pulsecore {

Deinterleaver::Deinterleaver(double sample_rate_hz, double pri_tolerance_seconds)
    : sample_rate_hz_(sample_rate_hz), pri_tolerance_seconds_(pri_tolerance_seconds) {}

void Deinterleaver::Process(const pulse::PulseEventBatch& batch, pulse::DeinterleaveSummary* out) {
  // Columnar events (see pulse.proto): entry e across the parallel
  // arrays is one pulse.
  const int n_events = batch.start_sample_size();
  for (int e = 0; e < n_events; ++e) {
    const uint64_t start_sample = batch.start_sample(e);
    const double pulse_time = static_cast<double>(start_sample) / sample_rate_hz_;

    // Prefer the closest track whose predicted next-pulse time falls
    // within tolerance; fall back to a track that only has one pulse so
    // far (no PRI to test against yet) if no established track matches.
    Track* best_match = nullptr;
    double best_diff = pri_tolerance_seconds_;
    Track* seed_match = nullptr;

    for (Track& track : tracks_) {
      if (track.pri_count > 0) {
        const double last_time = static_cast<double>(track.last_start_sample) / sample_rate_hz_;
        const double predicted =
            last_time + (track.pri_sum_seconds / static_cast<double>(track.pri_count));
        const double diff = std::abs(predicted - pulse_time);
        if (diff <= best_diff) {
          best_diff = diff;
          best_match = &track;
        }
      } else if (track.pulse_count == 1 && seed_match == nullptr) {
        seed_match = &track;
      }
    }

    Track* target = best_match != nullptr ? best_match : seed_match;
    if (target == nullptr) {
      tracks_.push_back(Track{next_track_id_++});
      target = &tracks_.back();
    }

    if (target->pulse_count > 0) {
      const double last_time = static_cast<double>(target->last_start_sample) / sample_rate_hz_;
      target->pri_sum_seconds += (pulse_time - last_time);
      ++target->pri_count;
    }
    target->last_start_sample = start_sample;
    target->peak_sum += batch.peak_amplitude(e);
    ++target->pulse_count;
  }

  out->Clear();
  for (const Track& track : tracks_) {
    pulse::EmitterTrack* t = out->add_tracks();
    t->set_track_id(track.track_id);
    t->set_pulse_count(track.pulse_count);
    t->set_estimated_pri_seconds(
        track.pri_count > 0 ? track.pri_sum_seconds / static_cast<double>(track.pri_count) : 0.0);
    t->set_mean_peak_amplitude(
        track.pulse_count > 0 ? track.peak_sum / static_cast<double>(track.pulse_count) : 0.0);
  }
}

}  // namespace pulsecore
